### DISCUSSION TITLE
How do you convert String to Id<T>?

### DISCUSSION THREAD

**Discussion by codegod100** [tokio-rs/toasty#24] [Q&A] [answered]
Trying to `find_by_id` but all the example have existing objects to call id on

**Comment by carllerche** [comment]
We should add a `from_string` method on it! You want to try to do it?

**Comment by codegod100** [comment]
Right now I'm using the following code
```
    let id = toasty_core::stmt::Id::from_string(ModelId(1), input.id);
    let id: Id<Bookmark> = Id::from_untyped(id);
    let bookmark = Bookmark::find_by_id(id).get(&state.db).await?;
```

I hard coded the model id because that is created via codegen and there's no accessor method that I could find. I have an idea of what I would do to add one though.

**Comment by carllerche** [comment]
I have some meetings now, I will get back to this after.

I do think it might be worth exploring whether' ID` as a concept (an opaque, model-tied identifier) holds its weight as such

Also, because query methods take `impl IntoExpr<_>` we probably could "just" implement that trait for `Id` and you could avoid the conversion.

**Reply by codegod100** [reply]
>we probably could "just" implement that trait for Id

Another thing that's kinda confusing is there is `toasty_core::stmt::id::Id` and `toasty::stmt::id::Id` so I'm not even sure which `Id` you're talking about

What is the design goal for having the Id value be typed like it is? 

Ostensibly, I want to say just use strings or bytes for all id values, but I'm not sure of the caveats of that.

**Reply by carllerche** [reply]
Some databases tend to use integers (auto-incrementing). `toasty_core::stmt::Id` is the "internal" version, and `toasty` is the user-facing version.

The general idea is that Both User and Todo types will have an identifier (whether string, integer, etc.), and the types should not be comparable. For example, it doesn't make sense to do `User::find_by_id(&todo.id)`. So, I typed them in a way that wouldn't be possible.

The big question is whether or not the friction of having a unique `id` type outweighs the win.

I will try to make your original example work, though.

**Reply by codegod100** [reply]
Is there a way to get the ModelId from the model? That seems to me the biggest blocker in writing a function that converts a string to the proper Id type.

**Reply by carllerche** [reply, marked as answer]
It is from the type: `M::Id` usually. I submitted a PR here: https://github.com/tokio-rs/toasty/pull/29
